kolibri/auth/test/test_collection_and_role_models.py

```diff
@@ -1,56 +1,72 @@
 
         self.cr = Classroom.objects.create(parent=self.facility)
-        self.cr.add_coach(user2)
+        self.cr.add_coach(classroom_coach)
 
         self.lg = LearnerGroup.objects.create(parent=self.cr)
-        self.lg.add_learner(user1)
+        self.lg.add_learner(learner)
 
     def test_remove_learner(self):
-        self.assertEqual(Role.objects.filter(user=self.user1, kind=Role.KIND_LEARNER, collection=self.lg).count(), 1)
-        self.lg.remove_learner(self.user1)
-        self.assertEqual(Role.objects.filter(user=self.user1, kind=Role.KIND_LEARNER, collection=self.lg).count(), 0)
+        self.assertEqual(Membership.objects.filter(user=self.learner, collection=self.lg).count(), 1)
+        self.assertTrue(self.lg.remove_learner(self.learner))
+        self.assertEqual(Membership.objects.filter(user=self.learner, collection=self.lg).count(), 0)
+        self.assertFalse(self.lg.remove_learner(self.learner))  # if it doesn't exist, removal returns False
 
     def test_remove_coach(self):
-        self.assertEqual(Role.objects.filter(user=self.user2, kind=Role.KIND_COACH, collection=self.cr).count(), 1)
-        self.cr.remove_coach(self.user2)
-        self.assertEqual(Role.objects.filter(user=self.user2, kind=Role.KIND_COACH, collection=self.cr).count(), 0)
+        self.assertEqual(Role.objects.filter(user=self.classroom_coach, kind=role_kinds.COACH, collection=self.cr).count(), 1)
+        self.assertTrue(self.cr.remove_coach(self.classroom_coach))
+        self.assertEqual(Role.objects.filter(user=self.classroom_coach, kind=role_kinds.COACH, collection=self.cr).count(), 0)
+        self.assertFalse(self.cr.remove_coach(self.classroom_coach))  # if it doesn't exist, removal returns False
 
     def test_remove_admin(self):
-        self.assertEqual(Role.objects.filter(user=self.user3, kind=Role.KIND_ADMIN, collection=self.facility).count(), 1)
-        self.facility.remove_admin(self.user3)
-        self.assertEqual(Role.objects.filter(user=self.user3, kind=Role.KIND_ADMIN, collection=self.facility).count(), 0)
+        self.assertEqual(Role.objects.filter(user=self.facility_admin, kind=role_kinds.ADMIN, collection=self.facility).count(), 1)
+        self.assertTrue(self.facility.remove_admin(self.facility_admin))
+        self.assertEqual(Role.objects.filter(user=self.facility_admin, kind=role_kinds.ADMIN, collection=self.facility).count(), 0)
+        self.assertFalse(self.facility.remove_admin(self.facility_admin))  # if it doesn't exist, removal returns False
+
+    def test_remove_nonexistent_role(self):
+        self.assertFalse(self.facility.remove_admin(self.learner))
+        self.assertFalse(self.cr.remove_coach(self.learner))
+
+    def test_remove_indirect_admin_role(self):
+        """ Trying to remove the admin role for a a Facility admin from a descendent classroom doesn't actually remove anything. """
+        self.assertFalse(self.cr.remove_admin(self.facility_admin))
```

Comment:
Seems like there's a distinction to be made between failing to remove a role because it doesn't exist vs failing to remove a role because somehow the request to do so doesn't make sense. Consider the following snippets, which might appear in a view for user management:

```python
if classroom.remove_admin(facility_admin):
    # The role was successfully removed
    return SuccessResponse("Role removed!")
else:
    return FailureResponse("Unable to remove role!")  # No indication as to why
```

vs

```python
try:
    removed = classroom.remove_admin(facility_admin)
except KolibriCoolHierarchyError:
    return FailureResponse("Can't remove, user possesses role from parent Facility!")
else:
    if removed:
        return SuccessResponse("Role removed!")
    else:
        return FailureResponse("User does not possess that role!")
```

Is it useful to make such a distinction in your opinion?